### PR TITLE
Improved wrong syntax popup.

### LINKED ISSUE #4455: Improve box dialog when opening_hours is incorrect in advanced mode
Scenario:
- Start OM
- Select POI on map
- Open editor -> open opening_hours IHM
- Select advanced mode
- Fill field with wrong syntax -> pop-up `fix error` appears

**Describe the solution you would like**
- Improve pop-up
- Title: Incorrect syntax
- Text in pop-up: Your opening_hours doesn't respect syntax, please check examples of values or copy paste your opening_hours on [YoHours](https://projets.pavie.info/yohours/)

https://github.com/organicmaps/organicmaps/blob/4a66bda647900bde3b02a9c08262632e1cf4fb50/android/src/app/organicmaps/editor/TimetableContainerFragment.java#L155

### PR DESCRIPTION
**Pull request by n1snt**
### Fixes #4455 issue

Before             |  After
:-------------------------:|:-------------------------:
 |

### CODE REVIEW

**Review by biodranik — CHANGES_REQUESTED** [review]
Thanks for the try! I've left more details on how this issue should be fixed here: https://github.com/organicmaps/organicmaps/issues/4455#issuecomment-1438207264

Please read it, and update your code/Pull Request with an appropriate solution.

**Comment by n1snt** [comment]
Sure will have a look at it.

**Comment by n1snt** [comment]
Error             |  Normal
:-------------------------:|:-------------------------:
image |  image

It now shows an error on edittext, greys out the save button & disables it if the input is not in the correct format.

File: android/src/app/organicmaps/editor/EditorHostFragment.java [hunk 1]
```diff
@@ -309,7 +309,7 @@ public void onClick(View v)
         else
         {
           // TODO (yunikkk) correct translation
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/EditorHostFragment.java:311]: Why the todo is not fixed? Why the existing string is not updated in strings.txt (see docs/TRANSLATIONS.md) and a new string is created instead?

**Review by vng — COMMENTED** [review]
Add new strings to the ./data/strings.txt and request for translations ..

**Comment by n1snt** [comment]
Pushed all the local changes. I am a bit confused about where I should be adding the strings

**Comment by Jean-BaptisteC** [comment]
Doing 2 commits:
- First commit with your changes on Java, remove `Correct mistake` string in https://github.com/organicmaps/organicmaps/blob/master/data/strings/strings.txt and add `format not valid` string in this same file (More information [here](https://github.com/organicmaps/organicmaps/blob/master/docs/TRANSLATIONS.md))
- Second commit with localization files regenerated

**Comment by n1snt** [comment]
Cannot remove the Correct mistake string as it's being used in another instance.
Have followed the instructions & regenerated the strings.Correct mistake

**Review by Jean-BaptisteC — APPROVED** [review]
✅ Pixel 6 - Android 13
Error works properly

Please sign your commits like explain [here](https://github.com/organicmaps/organicmaps/blob/master/docs/DCO.md)

File: android/res/values/strings.xml [hunk 2]
```diff
@@ -409,6 +409,7 @@
 	<string name="editor_hours_closed">Non-Business Hours</string>
 	<string name="editor_example_values">Example Values</string>
 	<string name="editor_correct_mistake">Correct mistake</string>
+	<string name="editor_invalid_format">Format not valid!</string>
```
Comment (biodranik) [inline android/res/values/strings.xml:411]: This change should not be done in this commit. It should be autogenerated by the script.

File: data/strings/strings.txt [hunk 3]
```diff
@@ -10021,6 +10021,10 @@
     zh-Hans = 示例值
     zh-Hant = 範例值
 
+  [editor_invalid_format]
```
Comment (biodranik) [inline data/strings/strings.txt:10024]: What is wrong with the existing translation editor_correct_mistake ? Why it can't be reused here?

File: android/src/app/organicmaps/editor/EditorHostFragment.java [hunk 4]
```diff
@@ -301,16 +301,8 @@ public void onClick(View v)
...
-        {
-          Editor.nativeSetOpeningHours(timetables);
-          editMapObject();
-        }
-        else
-        {
-          // TODO (yunikkk) correct translation
-          showMistakeDialog(R.string.editor_correct_mistake);
-        }
+        Editor.nativeSetOpeningHours(timetables);
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/EditorHostFragment.java:304]: This change should be squashed into the first commit.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 5]
```diff
@@ -137,7 +143,21 @@ private static void setTextChangedListener(@Nullable EditText input,
       public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
 
       @Override
-      public void onTextChanged(CharSequence s, int start, int before, int count) {}
+      public void onTextChanged(CharSequence s, int start, int before, int count) {
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:146]: Please follow our [code formatting rules](https://github.com/organicmaps/organicmaps/blob/master/docs/JAVA_STYLE.md).

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 6]
```diff
@@ -137,7 +143,21 @@ private static void setTextChangedListener(@Nullable EditText input,
...
+        // Check if valid
+        // If not valid
+        // show error on textview
+        // Disable and grey out save button
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:150]: This comment doesn't look good. Please use clear English sentences where necessary.

**Review by biodranik — CHANGES_REQUESTED** [review]

File: data/strings/strings.txt (hunk 3 above)
Comment (n1snt) [reply data/strings/strings.txt:10024]: `editor_correct_mistake` is not descriptive of what's wrong & it's being used in a few other instances in the codebase, so it cannot be removed.

Comment (biodranik) [reply data/strings/strings.txt:10024]: "Format not valid" is even worse: not every non-developer user can realize what "format" is. Considering that the error message should appear immediately after a typed symbol, it should be clear to a user what went wrong. And we're talking about the "Advanced" mode, right?

Comment (n1snt) [reply data/strings/strings.txt:10024]: `Correct mistake` only signifies user that there is a mistake.
`Format is not valid!` lets, users know that the mistake they are making is the wrong format.
This is the reason I went on with changing the string in this instance.
This error will only be displayed in advanced mode.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 6 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java]: Removed.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 5 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java]: Formatted using the formatting rules.

File: android/src/app/organicmaps/editor/EditorHostFragment.java (hunk 4 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/EditorHostFragment.java:304]: Squashed.

**Comment by n1snt** [comment]
Signed the commits & made all the requested changes. 👍

**Comment by biodranik** [comment]
1. Please make your code changes in a separate commit, _without_ any unnecessary formatting or string changes.
2. If you introduce a new translation string, commit it in the data/strings/strings.txt file in the next commit with "[strings] Error message for invalid opening hours" title.
3. Then regenerate translations with tools/unix/generate_localizations.sh in the last "[strings] Regenerated" commit, like it is done in other repository's commits (check repo history).

**Comment by Jean-BaptisteC** [comment]
Please remove changes about Twine

**Comment by n1snt** [comment]
@Jean-BaptisteC Done

**Review by biodranik — CHANGES_REQUESTED** [review]
After thorough testing, it looks like there is no need to add any additional error messages in the advanced opening hours mode.

Let's simplify it and do it like on iOS: if the wrong value is typed, then it can not be saved (the Done/OK button is greyed out and can't be pressed).

Those users who entered the Advanced mode should already know what they are doing. Or they can check the example/hint on the same screen to avoid mistakes.

**Comment by n1snt** [comment]
Made the requested changes @biodranik 👍

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 7]
```diff
@@ -22,12 +23,15 @@
 import app.organicmaps.util.UiUtils;
 
 public class AdvancedTimetableFragment extends BaseMwmFragment
-                                       implements View.OnClickListener, TimetableProvider
+    implements View.OnClickListener, TimetableProvider
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:26]: Please do not introduce unnecessary formatting changes.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 8]
```diff
@@ -8,12 +9,12 @@
 import android.view.ViewGroup;
 import android.webkit.WebView;
 import android.widget.EditText;
+import android.widget.ImageView;
 import android.widget.TextView;
 
 import androidx.annotation.DrawableRes;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
-
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:16]: Please do not introduce unnecessary changes.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 9]
```diff
@@ -85,7 +90,7 @@ private void showExample(boolean show)
   private void setExampleDrawables(@DrawableRes int left, @DrawableRes int right)
   {
     mExamplesTitle.setCompoundDrawablesRelativeWithIntrinsicBounds(Graphics.tint(requireActivity(), left, R.attr.colorAccent), null,
-                                                           Graphics.tint(requireActivity(), right, R.attr.colorAccent), null);
+                                                                   Graphics.tint(requireActivity(), right, R.attr.colorAccent), null);
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:93]: ditto

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 10]
```diff
@@ -125,19 +130,30 @@ void setTimetableChangedListener(@NonNull TimetableChangedListener listener)
...
   }
 
-  private static void setTextChangedListener(@Nullable EditText input,
-                                             @Nullable TimetableChangedListener listener)
+  private void setTextChangedListener(@Nullable EditText input,
+                                      @Nullable TimetableChangedListener listener)
   {
     if (input == null || listener == null)
       return;
-
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:133]: ditto

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 11]
```diff
@@ -125,19 +130,30 @@ void setTimetableChangedListener(@NonNull TimetableChangedListener listener)
...
     input.addTextChangedListener(new TextWatcher()
     {
       @Override
       public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
 
       @Override
-      public void onTextChanged(CharSequence s, int start, int before, int count) {}
+      public void onTextChanged(CharSequence s, int start, int before, int count)
+      {
+        boolean isValidFormat = OpeningHours.nativeIsTimetableStringValid(s.toString());
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:146]:
```suggestion
        final boolean isValidFormat = OpeningHours.nativeIsTimetableStringValid(s.toString());
```

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 12]
```diff
@@ -22,12 +23,15 @@
...
 public class AdvancedTimetableFragment extends BaseMwmFragment
-                                       implements View.OnClickListener, TimetableProvider
+    implements View.OnClickListener, TimetableProvider
 {
   private boolean mIsExampleShown;
   private EditText mInput;
   private WebView mExample;
   private TextView mExamplesTitle;
+
+  private ImageView mSaveButton;
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:33]:
```suggestion
  private ImageView mSaveButton;
```

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 13]
```diff
@@ -125,19 +130,30 @@ void setTimetableChangedListener(@NonNull TimetableChangedListener listener)
...
+        boolean isValidFormat = OpeningHours.nativeIsTimetableStringValid(s.toString());
+        mSaveButton.setEnabled(isValidFormat);
+        if (!isValidFormat)
+        {
+          mSaveButton.setColorFilter(Color.argb(150, 200, 200, 200));
+        }
+        else
+        {
+          mSaveButton.setColorFilter(null);
+        }
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:155]:
```suggestion
        mSaveButton.setColorFilter(isValidFormat ? null : Color.argb(150, 200, 200, 200));
```

**Review by biodranik — COMMENTED** [review]
Does it work well?

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 13 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:155]: For some reason, this change makes the app crash while launching the Advanced timetable.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 12 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java]: Done

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 11 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java]: Done.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 10 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java]: Done

**Comment by n1snt** [comment]
@biodranik Made all the requested changes 👍

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 13 above)
Comment (biodranik) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:155]: We are engineers, there is no "some reason", it should be a clear error message ;-)

Comment (biodranik) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:155]: Isn't the code I proposed identical to yours?

Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:155]: It throws this error:
```
java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.Integer.intValue()' on a null object reference
```
If I replace the null with Color.TRANSPARENT like this It does not crash: 
```
mSaveButton.setColorFilter(isValidFormat ? Color.TRANSPARENT : Color.argb(150, 200, 200, 200));
```

Comment (biodranik) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:155]: It is not clear why it works in your version because there is no setColorFilter(int) method.

On the other side, the button can be changed in the future, and hardcoding here the color value doesn't look like a flexible solution. It would be great to have only enable/disable call in the code, and specify colors for enabled/disabled state in resources, for example, like it is mentioned here: https://stackoverflow.com/a/33406917

Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:156]: The link you mentioned used textColor which cannot be used with our implementation since we are using an imageview as a button.
But there is `app:tint` which does the same thing.
I just updated my implementation of it have a look at it & let me know if there is anything to change there.
Here are some screenshots from the new implementation (changed the shade of grey this one looks much better in my opinion than the older one).
Enabled             |  Disabled
:-------------------------:|:-------------------------:
image | image

_Will sign & rebase the commits if you approve of this implementation :)_

File: android/res/color/image_view.xml [hunk 14]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+  <item android:color="@color/white_primary"
+    android:state_enabled="true"/>
+
+  <item android:color="@color/white_54"
+    android:state_enabled="false"/>
```
Comment (biodranik) [inline android/res/color/image_view.xml:7]:
```suggestion
  <item android:state_enabled="true"
    android:color="@color/white_primary"/>

  <item android:state_enabled="false"
    android:color="@color/white_54"/>
```

File: android/res/layout/fragment_editor_host.xml [hunk 15]
```diff
@@ -27,6 +28,7 @@
         tools:visibility="gone"/>
 
       <ImageView
+        app:tint="@color/image_view"
```
Comment (biodranik) [inline android/res/layout/fragment_editor_host.xml:31]: android:tint should also work here, right?

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 16]
```diff
@@ -65,6 +68,7 @@ private void initViews(View view)
     mExamplesTitle = view.findViewById(R.id.tv__examples_title);
     setExampleDrawables(R.drawable.ic_type_text, R.drawable.ic_expand_more);
     setTextChangedListener(mInput, mListener);
+    mSaveButton = getParentFragment().getParentFragment().getView().findViewById(R.id.save);
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:71]: Should this variable be defined/retrieved locally in the function and passed to the lambda instead of storing it as a class member?

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 17]
```diff
@@ -137,7 +141,21 @@ private static void setTextChangedListener(@Nullable EditText input,
       public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
 
       @Override
-      public void onTextChanged(CharSequence s, int start, int before, int count) {}
+      public void onTextChanged(CharSequence s, int start, int before, int count)
+      {
+        final boolean isValidFormat = OpeningHours.nativeIsTimetableStringValid(s.toString());
+        mSaveButton.setEnabled(isValidFormat);
+        /*
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:148]: Unused code should be removed.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java [hunk 18]
```diff
@@ -137,7 +141,21 @@ private static void setTextChangedListener(@Nullable EditText input,
       public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
 
       @Override
-      public void onTextChanged(CharSequence s, int start, int before, int count) {}
+      public void onTextChanged(CharSequence s, int start, int before, int count)
+      {
+        final boolean isValidFormat = OpeningHours.nativeIsTimetableStringValid(s.toString());
+        mSaveButton.setEnabled(isValidFormat);
```
Comment (biodranik) [inline android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:147]:
```suggestion
        saveButton.setEnabled(OpeningHours.nativeIsTimetableStringValid(s.toString()));
```

**Review by biodranik — COMMENTED** [review]
Did you try to edit some complex string? Does it work fast? Is there any delay when user types the text?

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 16 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:71]: If we move it to onTextChanged wouldn't it make the code slower? 
Because each time the text would change it would call the above line.

File: android/res/layout/fragment_editor_host.xml (hunk 15 above)
Comment (n1snt) [reply android/res/layout/fragment_editor_host.xml:31]: When I tried using android tint IDE was warning to use app:tint specifically.

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 17 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java]: Done

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 18 above)
Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java]: Done

File: android/res/color/image_view.xml (hunk 14 above)
Comment (n1snt) [reply android/res/color/image_view.xml]: Done

File: android/src/app/organicmaps/editor/AdvancedTimetableFragment.java (hunk 16 above)
Comment (biodranik) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:71]: I did not write to move it into onTextChanged. I wrote that it can be defined in the function (`setTextChangedListener`) and then passed into the onTextChanged. The variable is not needed outside of the lambda scope.

Comment (n1snt) [reply android/src/app/organicmaps/editor/AdvancedTimetableFragment.java:71]: setTextChangedListener() cannot be referenced from a static context so we can move it into the setTextChangedListener()  function but then we would need to make it non-static.

**Review by vng — APPROVED** [review]
Anything else here?

**Review by biodranik — APPROVED** [review]
That's it, thanks!
